kafka.go

```diff
@@ -0,0 +1,390 @@
+// Copyright (c) 2019 AccelByte Inc. All Rights Reserved.
+// This is licensed software from AccelByte Inc, for limitations
+// and restrictions contact your company contract manager.
+
+package messagebus
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/Shopify/sarama"
+	"github.com/sirupsen/logrus"
+
+	"time"
+)
+
+var (
+	rtoError = errors.New("request time out in publish with response")
+)
+
+var subscribeMap map[string]map[string]func(message *Message, err error)
+
+var publishResponseTimeout = 10000
+
+// KafkaClient wraps client's functionality for Kafka
+type KafkaClient struct {
+	asyncProducer sarama.AsyncProducer
+	syncProducer  sarama.SyncProducer
+	consumer      sarama.Consumer
+	broker        *sarama.Broker
+}
+
+// PublishBuilderEntry defines data structure to use with Kafka client
+type PublishBuilderEntry struct {
+	PublishBuilder
+
+	encoded []byte
+	err     error
+}
+
+func (publishBuilderEntry *PublishBuilderEntry) ensureEncoded() {
+	if publishBuilderEntry.encoded == nil && publishBuilderEntry.err == nil {
+
+		if publishBuilderEntry.traceId == "" {
+			publishBuilderEntry.traceId = generateID()
+		}
+
+		if publishBuilderEntry.messageId == "" {
+			publishBuilderEntry.messageId = generateID()
+		}
+
+		publishBuilderEntry.encoded, publishBuilderEntry.err = json.Marshal(Message{
+			Message:     publishBuilderEntry.message,
+			MessageType: publishBuilderEntry.messageType,
+			Service:     publishBuilderEntry.service,
+			TraceId:     publishBuilderEntry.traceId,
+			MessageId:   publishBuilderEntry.messageId,
+		})
+	}
+}
+
+// Encode PublishBuilder into array of bytes
+func (publishBuilderEntry *PublishBuilderEntry) Encode() ([]byte, error) {
+	publishBuilderEntry.ensureEncoded()
+	return publishBuilderEntry.encoded, publishBuilderEntry.err
+}
+
+// Length returns size of encoded value
+func (publishBuilderEntry *PublishBuilderEntry) Length() int {
+	publishBuilderEntry.ensureEncoded()
+	return len(publishBuilderEntry.encoded)
+}
+
+// NewKafkaClient create a new instance of KafkaClient
+func NewKafkaClient(brokerList []string) (*KafkaClient, error) {
+	config := sarama.NewConfig()
+	config.Version = sarama.V2_1_0_0
+
+	// currently only support 1 message broker
+	broker := sarama.NewBroker(brokerList[0])
+	_ = broker.Open(config)
```

Comment:
is this an error that you silent? or something else?

Reply:
you're right, I will update it
thanks